### PR TITLE
Async tracebacks

### PR DESCRIPTION
**Pull request by dom96**
Removes a lot of the noise in async trace backs and provides some hints as well.

### CODE REVIEW

**Comment by Yardanico** [comment]
Wow, nice one!

**Comment by dom96** [comment]
@Yardanico Thanks :) Haven't seen you in IRC for a while, how have you been?

File: lib/pure/asyncfutures.nim [hunk 1]
```diff
@@ -217,17 +217,115 @@ proc `callback=`*[T](future: Future[T],
...
+  ## together with the index of the last match.
+  ##
+  ## .. code-block::nim
+  ##   doAssert(diff(@[1,2,42,123], @[1,2,123,678,21]) == (1, @[123,678,21]))
+  var lastIndex = 0
+  for i in 0..<min(a.len, b.len):
+    lastIndex = i
+    if a[i] != b[i]:
+      break
+  return (lastIndex, b[lastIndex .. ^1])
```
Comment (Araq) [inline lib/pure/asyncfutures.nim]: Expensive to return this seq copy which doesn't add any information.

File: lib/pure/asyncfutures.nim [hunk 2]
```diff
@@ -217,17 +217,115 @@ proc `callback=`*[T](future: Future[T],
...
+    if entry.procName.isNil:
+      # Start of a re-raise traceback which may contain more info.
+      # Find where the re-raised traceback ends.
+      assert entry.line == -10 # Signifies start of re-raise block.
+      var reRaiseEnd = i+1
+      while reRaiseEnd < entries.len and not entries[reRaiseEnd].procName.isNil:
+        reRaiseEnd.inc()
+      assert entries[reRaiseEnd].procName.isNil
+      assert entries[reRaiseEnd].line == -100 # Signifies end of re-raise block.
+      let reRaisedEntries = entries[i+1 .. reRaiseEnd-1]
```
Comment (Araq) [inline lib/pure/asyncfutures.nim]: Expensive.

File: lib/pure/asyncfutures.nim [hunk 3]
```diff
@@ -217,17 +217,115 @@ proc `callback=`*[T](future: Future[T],
...
+
+      let (lastIndex, remainder) = diff(result, reRaisedEntries)
+      # Insert all the entries after lastIndex.
+      for i in 0..<remainder.len:
+        result.insert(remainder[i], lastIndex+i)
```
Comment (Araq) [inline lib/pure/asyncfutures.nim]: Expensive.

File: lib/pure/asyncfutures.nim [hunk 4]
```diff
@@ -217,17 +217,115 @@ proc `callback=`*[T](future: Future[T],
...
+      continue
+
+    result.add(entry)
+    i.inc
+
+proc getHint(entry: StackTraceEntry): string =
+  ## We try to provide some hints about stack trace entries that the user
+  ## may not be familiar with, in particular calls inside the stdlib.
+  result = ""
+  let name = ($entry.procName).normalize()
```
Comment (Araq) [inline lib/pure/asyncfutures.nim]: Expensive.

File: lib/pure/asyncfutures.nim [hunk 5]
```diff
@@ -217,17 +217,115 @@ proc `callback=`*[T](future: Future[T],
...
+  case name
+  of "processpendingcallbacks":
```
Comment (Araq) [inline lib/pure/asyncfutures.nim]: Hacks.

**Comment by Araq** [comment]
I know it's probably not on the critical path but it's so inefficient that it pains me.

File: lib/pure/asyncfutures.nim [hunk 6]
```diff
@@ -264,17 +259,15 @@ proc getHint(entry: StackTraceEntry): string =
   ## We try to provide some hints about stack trace entries that the user
   ## may not be familiar with, in particular calls inside the stdlib.
   result = ""
-  let name = ($entry.procName).normalize()
-  case name
-  of "processpendingcallbacks":
-    if cmpIgnoreStyle($entry.filename, "asyncdispatch.nim") == 0:
+  if entry.procname == "processPendingCallbacks":
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: In what way is this better than a case statement?

File: lib/pure/asyncfutures.nim [hunk 7]
```diff
@@ -217,19 +217,15 @@ proc `callback=`*[T](future: Future[T],
   ## If future has already completed then ``cb`` will be called immediately.
   future.callback = proc () = cb(future)
 
-proc diff[T](a, b: seq[T]): (int, seq[T]) =
+proc diff[T](a, b: seq[T]; firstB, lastB: int): int =
   ## Iterates through both sequences until the items do not match,
-  ## returns the remainder of `b` after the last item that does not match
+  ## returns the remainder of `b[firstB..lastB]` after the last item that does not match
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: this doc string doesn't look right

**Comment by dom96** [comment]
Yeah, the fact that this isn't a critical path really limits how much speeding up this code gets you in terms of performance.

I'm happy for you to do it, as long as it doesn't obfuscate the code too much and the tests still pass. Both things, I'm sorry to say, I don't think your changes satisfy :)

**Comment by Araq** [comment]
> I'm happy for you to do it, as long as it doesn't obfuscate the code too much and the tests still pass. Both things, I'm sorry to say, I don't think your changes satisfy

Not **yet**. :P

**Comment by dom96** [comment]
You've obfuscated it even more... :(

**Comment by dom96** [comment]
Oh well. Let's see if it passes. Writing "'diff' is the wrong idea here." has no meaning for people reading this code though as 'diff' isn't in the code.

**Comment by Yardanico** [comment]
F*** mobile, sorry, mistap

**Pull request merged**